nexus/src/app/sled.rs

```diff
@@ -231,4 +233,166 @@ impl super::Nexus {
         self.db_datastore.service_upsert(opctx, service).await?;
         Ok(())
     }
+
+    // OPTE V2P mappings
+
+    /// Ensure that the necessary v2p mappings for an instance are created
+    pub async fn create_instance_v2p_mappings(
+        &self,
+        opctx: &OpContext,
+        instance_id: Uuid,
+    ) -> Result<(), Error> {
+        // For every sled that isn't the sled this instance was allocated to, create
+        // a virtual to physical mapping for each of this instance's NICs.
```

Comment:
I'm not sure about this approach. I get the point of a quick solution, but a more appropriate one doesn't seem like significantly more work. We're creating a bunch of irrelevant mappings. And more importantly, we've explicitly decided to defer VPC peering in the MVP. I don't think we _want_ to send mappings for instances in another VPC. You rightly point out that OPTE will drop such traffic, but I don't think we should bake that knowledge in here. I think it makes sense to try to send the mappings we explicitly want to allow.

As an alternative, what about including a function call here, something like `Nexus::sleds_for_vpc_mapping()`. That would currently select all sleds that are in the same VPC as the provided instance. In the future, we can update that to make more fine-grained decisions, taking into account peering for example. But we have an interface for selecting the "right" sleds, encapsulating what "right" means.

I believe the actual SQL query to derive this information would be pretty simple. It should look basically like this:

```sql
SELECT
    active_server_id
FROM
    (
        SELECT
            instance_id
        FROM
            network_interface
        WHERE
            vpc_id = $VPC_ID
            AND time_deleted IS NULL
        LIMIT
            1
    )
        AS inst
    INNER JOIN instance ON inst.instance_id = instance.id
```

That selects the ID of all instances from the NIC table sharing the same VPC. (You have that from the call to `derive_network_interface_info()`.) Then we join with the instance table on the primary key, and select the sled ID. It's not too much additional work to join on the sled table and pull the IP directly, too.

This is definitely no worse than we have now, in terms of atomicity, and I think it allows us to more faithfully represent the intent and evolve as we need finer-grained data.

Reply:
Oh, I should have added a big caveat! Diesel can be an absolute nightmare sometimes. The SQL query here is pretty easy, but it involves a subquery. That can sometimes be very hard to express in Diesel. I _think_ it should be straightforward here, but if we find that it can't be expressed easily in Diesel's DSL, then we should seriously reconsider. Hand-rolled queries are definitely much more work.

Reply:
It's not that it's impossible to do, but I'm anxious about TOCTOU (like Greg mentioned in another comment). Without something always running in the background that can apply corrections we run the risk of these mappings not being correct.

Reply:
I agree there is a risk of these mappings being incorrect, for example if an instance is moved before they can be applied to all sleds. I don't see how propagating them to sleds that don't need it solves that. As an example, if we propagate a mapping to some sled, but the instance moves, that sled now has the wrong mapping in any case, even if that's only for some period of time.

I had imagined we'd solve this mechanisms put forward in RFD 373 on reliable persistent workflows, such as [generation numbers](https://rfd.shared.oxide.computer/rfd/0373#_generation_numbers). We could compute the set of mappings at some point, along with generation number. If a sled receives a mapping that is prior to its existing generation, it may simply ignore it. It still seems like we want to compute the right set of data at each generation.

Reply:
@bnaecker: @gjcolombo and I discussed the broadcast method, and are convinced at this point that it seems safe. See the comment chain at https://github.com/oxidecomputer/omicron/pull/2536#discussion_r1133143006.

After a re-read, I *think* I grok the generation number approach outlined in RFD 373. At this point though I think I'd like to create a follow on issue for that work given that we're thinking the broadcast method is safe. Thoughts?